contracts/src/token/erc1155/extensions/supply.rs

```diff
@@ -0,0 +1,824 @@
+        approved: bool,
+    ) -> Result<(), erc1155::Error>;
+
+    /// Returns true if `operator` is approved to transfer `account`'s
+    /// tokens.
+    ///
+    /// Re-export of [`Erc1155::is_approved_for_all`].
+    #[allow(clippy::missing_errors_doc)]
+    fn is_approved_for_all(&self, account: Address, operator: Address) -> bool;
+
+    /// Transfers a `value` amount of tokens of type `id` from `from` to
+    /// `to`.
+    ///
+    /// Re-export of [`Erc1155::safe_transfer_from`].
+    #[allow(clippy::missing_errors_doc)]
+    fn safe_transfer_from(
+        &mut self,
+        from: Address,
+        to: Address,
+        id: U256,
+        value: U256,
+        data: Bytes,
+    ) -> Result<(), erc1155::Error>;
+
+    /// Batched version of [`IErc1155Supply::safe_transfer_from`].
+    ///
+    /// Re-export of [`Erc1155::safe_batch_transfer_from`].
+    #[allow(clippy::missing_errors_doc)]
+    fn safe_batch_transfer_from(
+        &mut self,
+        from: Address,
+        to: Address,
+        ids: Vec<U256>,
+        values: Vec<U256>,
+        data: Bytes,
+    ) -> Result<(), erc1155::Error>;
+}
+
+#[public]
+impl IErc1155Supply for Erc1155Supply {
+    type Error = erc1155::Error;
+
+    fn total_supply(&self, id: U256) -> U256 {
+        self._total_supply.get(id)
+    }
+
+    #[selector(name = "totalSupply")]
+    fn total_supply_all(&self) -> U256 {
+        *self._total_supply_all
+    }
+
+    fn exists(&self, id: U256) -> bool {
+        self.total_supply(id) > U256::ZERO
+    }
+
+    fn balance_of(&self, account: Address, id: U256) -> U256 {
+        self.erc1155.balance_of(account, id)
+    }
+
+    fn balance_of_batch(
+        &self,
+        accounts: Vec<Address>,
+        ids: Vec<U256>,
+    ) -> Result<Vec<U256>, erc1155::Error> {
+        self.erc1155.balance_of_batch(accounts, ids)
+    }
+
+    fn set_approval_for_all(
+        &mut self,
+        operator: Address,
+        approved: bool,
+    ) -> Result<(), erc1155::Error> {
+        self.erc1155.set_approval_for_all(operator, approved)
+    }
+
+    fn is_approved_for_all(&self, account: Address, operator: Address) -> bool {
+        self.erc1155.is_approved_for_all(account, operator)
+    }
+
+    fn safe_transfer_from(
+        &mut self,
+        from: Address,
+        to: Address,
+        id: U256,
+        value: U256,
+        data: Bytes,
+    ) -> Result<(), erc1155::Error> {
+        self.erc1155.authorize_transfer(from)?;
+        self.do_safe_transfer_from(from, to, vec![id], vec![value], &data)
+    }
+
+    fn safe_batch_transfer_from(
+        &mut self,
+        from: Address,
+        to: Address,
+        ids: Vec<U256>,
+        values: Vec<U256>,
+        data: Bytes,
+    ) -> Result<(), erc1155::Error> {
+        self.erc1155.authorize_transfer(from)?;
+        self.do_safe_transfer_from(from, to, ids, values, &data)
+    }
+}
+
+impl Erc1155Supply {
+    // Note: overriding `_update` requires reimplementing all of the functions
```

Comment:
In Rust we do not have Solidity like overriding... I believe we should remove this comment.